src/managed/Microsoft.Extensions.DependencyModel/StreamBufferWriter.Netstandard2_0.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+
+namespace Microsoft.Extensions.DependencyModel
+{
+    internal struct StreamBufferWriter : IBufferWriter<byte>, IDisposable
+    {
+        private Stream _stream;
+        private byte[] _rentedBuffer;
+
+        private const int MinimumBufferSize = 256;
+
+        public StreamBufferWriter(Stream stream, int bufferSize = MinimumBufferSize)
+        {
+            if (bufferSize <= 0)
+                throw new ArgumentException(nameof(bufferSize));
+
+            _stream = stream ?? throw new ArgumentNullException(nameof(stream));
+
+            _rentedBuffer = ArrayPool<byte>.Shared.Rent(bufferSize);
+        }
+
+        public void Advance(int count)
+        {
+            _stream.Write(_rentedBuffer, 0, count);
+            _rentedBuffer.AsSpan(0, count).Clear();
```

Comment:
Why is this clear needed?

Also, I'd have expected it to be an error condition if you advanced further than you'd requested memory for.  That doesn't need to happen / be tracked?

Reply:
You are right, the clear isn't needed, and yes, I need to do the input validation here.

I spent some time discussing with @davidfowl around a stream<->IBufferWriter impl, and the conclusion was that we shouldn't have such an adapter (and definitely not one that writes to the stream synchronously on calls to Advance).

Instead, we should buffer the whole output in memory (in some re-sizable array-backed IBufferWriter) and then copy the data to the stream asynchronously (using a `CopyToAsync(Stream stream)` method).

Something like the `ArrayFormatter`: https://github.com/dotnet/corefxlab/blob/87990483caf421669efb1f8f63f333b065d553b7/src/System.Text.Formatting/System/Text/Formatting/Formatters/ArrayFormatter.cs#L43